### PR TITLE
Using Masked inside Time

### PR DESCRIPTION
**Pull request by mhvk**
To do:
- [x] Make mask read-only (at least as gotten from `.mask`, perhaps also internally).
- [x] Add configuration option for having `np.ma.MaskedArray` always on output.
- [x] Add documentation (change-log entry, what's new, documentation about config)

@taldcroft - when I introduced `Masked`, I had hoped to soon use it to have masked `SkyCoord`, etc. Of course, things never go very fast. But even at the time I had made a start by seeing how things would work for `Time`, where it would replace using `np.nan` for `jd2`. This PR is to actually do that (as draft for now). This is triggered in part by #15230 which led back to #6509, a request to have an option for "NaT", ie., a marker for a bad time (as opposed to a masked one).

I remember vaguely that at the time you were rather hesitant to go this route, but do not really remember why, except that it probably involved backward compatibility! The approach taken here is that if `Time` gets `np.ma.MaskedArray` as input, that's also what it will give as output (reusing the `_shaped_like_input` method as before), otherwise it will be `Masked`. I think this is a reasonable option, though I think we should add a configuration option for those who want to just set a given type. If you think that makes sense, I'll add that.

Note that the PR consists of 4 commits all of which are working states, so perhaps useful to review independently:
1. Move to use `Masked` everywhere. Conversion upon inputs, all output as `Masked`. EDIT: also a small bug fix in `fitstime`, uncovered because now jd2 is no longer nan if an element is masked.
2. Since formats do not really have to know about masking any more, move the mask stuff to `Time` proper.
3. For outputs, use the masked class used for the inputs. *Note:* this is incomplete: if one does `t2 = t-t[0]`, then `t2` just uses `Masked` again. Probably should take `_masked_cls` from the left-most argument or so. Or maybe this should be an attribute on `info` and let that do the merging?!
4. Extra tests for all formats and a fix for a bug found as a result (can mix in 1 and 3 if desired).

One API change is that the `mask` is now writeable if masked values were passed in. This is consistent with how masked arrays generally work (but could be changed). One consequence of that is that the `masked` property can no longer be cached. (Obviously, this could be changed. Since the mask is always a copy of that from the inputs, we could write-protect it.)

Out of scope here:
1. Solving #6509 - though if we use `Masked` internally, we become free to use `np.nan` to indicate "NaT";
2. Allowing input of masked data where the underlying values do not get overridden (right now, one would have to initialize `t = Time(val.unmasked, ...)` and then set the mask after, with `t[val.mask] = np.ma.masked`).
3. Might be nice for `Masked(times, mask=...)` to work too.

Small trial:
```
from astropy.time import Time
from astropy.utils.masked import Masked
t = Time(Masked([50000., 50001, 50002], mask=[False, True, False]), format='mjd')
t
# <Time object: scale='utc' format='mjd' value=[50000.    ——— 50002.]>
t.isot
# MaskedNDArray(['1995-10-10T00:00:00.000',                       ———,
#                '1995-10-12T00:00:00.000'], dtype='<U23')
(t-t[0]).to('s')
# <MaskedQuantity [     0.,     ———, 172800.] s>

from astropy.table import QTable
QTable([t])
 
<QTable length=3>
  col0 
  Time 
-------
50000.0
    ———
50002.0
```

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by taldcroft** [comment]
@mhvk - you got it that my big concern would be about breaking existing code. The minimal impact to existing astropy tests is a good start, but some quick questions without really looking at the code yet:

- You mentioned that the `mask` is writeable, so would that cause problems for the caching of format conversions? (e.g. `tm.iso` is cached).
- To what extent is your `MaskedNDArray` API-compatible with the current numpy `MaskedArray`? What is the potential for incompatibility there?
- What about speed and memory performance changes?
- Does this fix #15230 ?

**Comment by mhvk** [comment]
> @mhvk - you got it that my big concern would be about breaking existing code. The minimal impact to existing astropy tests is a good start, but some quick questions without really looking at the code yet:
> 
>     * You mentioned that the `mask` is writeable, so would that cause problems for the caching of format conversions? (e.g. `tm.iso` is cached).

I do indeed worry about that, and have not checked. I do not have a strong opinion on whether the mask should be writeable or not, so we can stick more to the `Time`-is-almost-immutable case that we have and make the mask itself (or perpaps what is returned by `.mask`) read-only again. Indeed, I just realized that since the mask is shared between jd1 and jd2, making it read-only might give some extra protection against them somehow getting out of sync. 
 
>     * To what extent is your `MaskedNDArray` API-compatible with the current numpy `MaskedArray`? What is the potential for incompatibility there?

See https://docs.astropy.org/en/latest/utils/masked/index.html#utils-masked-vs-numpy-maskedarray

I *think* that if we implement the config item I suggested that enforces `np.ma.MaskedArray` output, users of `Time` can ensure there are no differences whatsoever.

>     * What about speed and memory performance changes?

I would expect a (slight) degradation in both cases, since there now is some overhead in propagating and storing the mask (latter only 1 byte for 16 bytes of data, though, since mask is shared). Given that actual time calculations are expensive, I think the time overhead will be minimal. Would be good if we still had functioning benchmarks! (`astropy-benchmarks` seems pretty defunct, and sadly `astropy.time` never made it on them). Of course, this only concerns the case that `Time` is actually masked!

>     * Does this fix [`TypeError` when comparing TimeDelta and a nan Quantity #15230](https://github.com/astropy/astropy/issues/15230) ?

Not directly. But it will enable using `np.nan` as a marker for NaT (though I think we could fix #15230 also by simply allowing `TimeDelta` to have non-finite numbers; unlike for `Time`, that might break just about nothing).

Note that the larger goal I still have in mind here is to add masking to `SkyCoord` (and everything below it), and potentially even modelling. Since `SkyCoord` uses `Quantity`, I will have to use `Masked` there. A bit similarly, I have been working on `Distribution` recently, to allow it to be used in representations, etc., and thus get "automatic" error propagation also for coordinates. To get that to work for `Time` needs an overhaul very similar to what I did here, so that `jd1` and `jd2` are no longer required to be base `ndarray`.

Anyway, am happy to make changes, but would like to be sure you think this is OK at least in principle...

**Comment by taldcroft** [comment]
@mhvk - in principle I am fine going ahead. In part this is because I think `SkyCoord` really needs masking and so doing things in a unified way is the right approach. The `nan` approach for `Time` was always meant as a "good-enough" approach in lieu of another mature solution. Since `Masked` has been out for some time now it should be solid enough.

Thanks for the work!

**Comment by mhvk** [comment]
OK, sounds good. I add two checks on top for things I should implement.

**Comment by mhvk** [comment]
@taldcroft - I pushed an update to this effort to use `Masked` inside `Time`, which includes updates to the documentation (plus changelog entry and what's-new), adds a configuration option to get `np.ma.MaskedArray` output, and makes the `.mask` read-only.

Compared to what I had before, the main change is that I kept the option simple: choose either `Masked` or `np.ma.MaskedArray` output, i.e., I dropped my original idea that the output should by default depend on what class was used for input. This mostly because propagating that information through operations seemed needlessly complex, but also just to "refuse the temptation to guess". 

This is fairly complete, but there are a few related things I'm wondering about; would be great to get your opinion!
1. Should `.masked` now just depend on whether the internal arrays are `Masked`, rather than on whether there is any element masked? When using `np.nan` we really didn't have much choice, but always having a mask when the data is `Masked` might help avoid surprises, that, e.g., one can only do `time.jd.filled(np.nan)` if any element is masked, even if one passed `Masked` input (but perhaps had no masked elements).
2. Related: should it be possible make `Time` masked? E.g., by setting `time.masked = True`. Or, perhaps nicer, by making `Masked(time, mask=...)` work? The latter may need a bit more thinking, but should be do-able (normally, one would have separate `MaskedTime` and `Time` classes, but I don't think it is best here and it should be possible to avoid it).
3. Related too: should `Time` gain a `.filled()` method and/or an `unmasked` property? (Latter could also be done by `time.masked = False`.)

Here, all could be done as follow-up, but I think at least the first point would be best addressed before 6.0, since otherwise people get another API change in 6.1. The other items just add features, so could be done in a later version too (but good to think through a bit what we want, since it would affect `SkyCoord` as well).

**Comment by mhvk** [comment]
p.s. I checked, and it turns out to be fairly trivial to get `Masked(time, mask=...)` and to even have a `MaskedTime = Masked(Time)` which takes a `mask` argument. One could imagine that this `MaskedTime` has the `unmasked` property and `filled()` method, while regular `Time` can have a mask, but does not have those "extras". In that case, we could leave the behaviour of `Time.masked` exactly as it is now, since one can do `ifinstance(possibly_masked_time, Masked)`.

p.s.2 The scheme I used to create `MaskedTime` would work trivially on other classes like `SkyCoord`!

**Comment by taldcroft** [comment]
This makes me think about a big picture question and overall API consistency. 

The current `Time` class can be masked and that is indicated by the `masked` attribute. Because `Time` is a container class it has the option of converting to a "masked" object in-place, and presumably the same would apply to `SkyCoord`. However, `Column` and `Quantity` are different and for those they require a distinct `Masked...` version.

If there will be a `MaskedTime` which is slight different from `Time` with masked JD values, then would a table join with missing values create a `Time` or a `MaskedTime`? Would one choice be preferred?

Writing it out like this makes me think that we should adhere to "one and only one way" to doing masking for each class at the expense of inconsistency in how masking is done.

That would imply answers to your questions (of course up for more discussion):

1. Should .masked now just depend on whether the internal arrays are Masked? Yes.
2. Should `Masked(time, mask=...)` work? Whether or not it works, astropy core should not document that as part of core astropy nor define or use a `MaskedTime` class anywhere.
3. Should Time gain a .filled() method and/or an unmasked property? Yes, those are both well defined even for an object that is not masked so the API can be consistent.

**Comment by mhvk** [comment]
Thanks, that's super helpful! I agree with your logic. Shall I add `.filled()` and `unmasked` to this PR, then?

One further question: it would seem sensible to allow setting `Time.masked` -- otherwise, to make an instance `masked` without setting any masked elements would require something a bit silly like doing `t[:0] = np.ma.masked`. Setting to `False` would then be equivalent to doing `.unmasked` inplace.

As for working with `Masked`, I made a few trials and it is possible to recognize `Time` as a `Masked` instance if it is in fact `Masked`, which I think will be useful. The `MaskedTime` class is easily made into a pure factory class, which just adds a `mask` argument to `Time`, but produces a `Time` instance (for which `instance(t, MaskedTime)` will work as long as `t.masked` holds. But that is for another PR -- I think much of the code can live on a new `MaskableShapedLikeNDarray` class, which can then be used by `SkyCoord`, etc., as well.

**Comment by taldcroft** [comment]
@mhvk - About the `.masked` attribute. Seeing your question makes me think again and wonder if there will be any internal or downstream issues with the subtle API change in what `.masked` implies (has some masked items vs *might* have masked items). I took a quick look through astropy time and table and nothing jumped out.

In any case, I suppose that `.masked` could be settable, but the more important question is whether that should ever be necessary. I would expect that doing `tm[2] = np.ma.masked` for an unmasked `Time` would auto-convert the `_time` object to Masked and that would be it. In other words why would you want to set `.masked` without actually masking any item?

Conceptually this gets to the original implementation and analogy with Pandas that from the user perspective there doesn't need to be much (if any) distinction between masked or unmasked. This is the benefit of having a container class (or using sentinel values).

In the other direction, doing `tm.masked = False` will have the consequence of unmasking values in-place, which might be surprising to some users. The explicit way to write this would be `tm = tm.unmasked`.

Upshot: I'm on the fence about whether to *allow* setting `.masked`, but either way I don't think we should *document* this is a feature of the API.

**Comment by mhvk** [comment]
OK, why don't I leave settable `masked` out of this PR -- it is something that can be added later anyway.

**Comment by mhvk** [comment]
OK, I pushed the further commits that make `time.masked` reflect whether or not the internal `jd1` is `Masked`.

**Comment by mhvk** [comment]
Note that the test on scalars uncovered a small bug in `Masked` (bca63dc5514521b0f89bf2fe17bc8ca5838c3b39) - I'll make a separate PR with that, since that probably should still be backported.

**Comment by taldcroft** [comment]
Here is where we come back to the fundamental issue of masked vs missing. This is surprising to me because of the implication that the data are valid but simply masked:
```
>>> dates = Masked(["2000-01-01", "2003-12-30"], mask=[True, False])
>>> print(Time(dates, format="iso", out_subfmt="date").unmasked)
['2003-12-30' '2003-12-30']
```
Most people would probably expect `['2000-01-01', '2003-12-30']`. But of course the code should also work for the example of *missing* or *bad* data (and indeed it does work but with a result that might surprise some people):
```
>>> dates = Masked(["", "2003-12-30"], mask=[True, False])
>>> print(Time(dates, format="iso", out_subfmt="date").unmasked)
['2003-12-30' '2003-12-30']
```

I wonder if we could address this ambiguity with this:
1. Define a cached property for each format class that provides a default value which is equivalent to "2000-01-01".
2. Use that default value consistently to replace any masked / missing values.
3. Document that new property and in particular make it clear for Time that any input masked values are always replaced with the default value.

Note that these issues are not necessarily new with this PR, but are somehow more obvious to me now.

**Comment by taldcroft** [comment]
This fails:
```
Time(Masked(["2000-01-01"], mask=[True]), format="iso")
```
Again not new but probably a good time to fix it, presumably with the above approach.

**Comment by mhvk** [comment]
@taldcroft - yes, it is annoying that input with masked values and setting a mask after are different. Right now, I tried to keep things as much the same as they were, but of course previously masked items effectively became NaN in output, while now one can see the good-enough value used. 

Ideally, we'd combine this with having not-a-time (NaT), but maybe your idea of effectively having a default `fill_value` can work too. Part of the problem is that the formats never had to deal with masking themselves, so if we introduce that, it is possible that would break custom formats.

I need to think about this more, but unfortunately have this research proposal that is due Nov 1st and which I'm quite behind schedule for...  I'm not sure what is best, try to put something like what I have now in, and then improve on it for 6.1 (which I think is allowed), or wait for 7.0 (since this PR is a clear API change).

**Comment by mhvk** [comment]
Super! I did think a bit about what state we would like to reach. I think
1. For all-masked, all we can do is rely on the user having taken care that the underlying data is not broken (a bit like for `np.ma.MaskedArray` assuming that `fill_value` is fine).
2. In the ideal world, we'd replace only broken underlying data with NaT, or, failing support for that, with some default.

But for this PR, it would definitely be nice to keep things simpler...

**Comment by taldcroft** [comment]
@mhvk - here is a branch I made, starting with rebasing on main. There was one conflict in the masked subpackage I wasn't sure about, but I basically chose incoming (#15451) over the conflicting changes in this PR related to formatting.

https://github.com/astropy/astropy/compare/main...taldcroft:astropy:pr-mhvk-15231-rebase

You can see the commits I've made since then.

> For all-masked, all we can do is rely on the user having taken care that the underlying data is not broken

This PR handles that case just fine. I'm in the process of writing more tests but figure I'd get this to you first.

> In the ideal world, we'd replace only broken underlying data

I think this is fragile and too complex. The main issue stems from guessing the format, where it gets messy if you are individually checking the masked values (for every format) for compliance with the format and replacing only the failures. Everything is possible in code but at a cost of complexity (both in the code and more importantly in the user documentation *explaining* how it works).

Instead just a short simple rule that any masked values on input get replaced with the equivalent of 2000-01-01.

**Comment by taldcroft** [comment]
Here are just my changes (basically what my PR on your PR would look like after you rebase).
https://github.com/taldcroft/astropy/compare/1b8a553c7612042ae034107e69d7f9cdd0de9592...1ba39306d4011a88e47ba9e048853e62e2f93a6d

**Comment by mhvk** [comment]
@taldcroft - had a few minutes and couldn't resist looking... Those changes are brilliant! I especially like how you avoided having to change any user `TimeFormat` classes, and made things substantially simpler in the process. Do you want to just force-push your rebased plus improved stuff to this branch?

**Comment by mhvk** [comment]
p.s. I'm super fine with a standard default value instead of trying to be clever about the input, though perhaps pick "J2000" (which is noon on 2000-01-01).

**Comment by taldcroft** [comment]
@mhvk - done. 🤞 I didn't mess anything with the force push, you best keep a copy of your local branch before pulling this!

File: astropy/time/core.py [hunk 1]
```diff
@@ -593,6 +607,8 @@ def _get_time_fmt(self, val, val2, format, scale, precision, in_subfmt, out_subf
         problems = {}
         for name, cls in formats:
             try:
+                if np.any(mask):
+                    val, val2 = cls._fill_masked_values(val, val2, mask, in_subfmt)
```
Comment (mhvk) [inline astropy/time/core.py:611]: This seems a bit risky, as now a badly written `_fill_masked_values` for one format can cause problems for the next one; I'll revise.

**Review by mhvk — COMMENTED** [review]
@taldcroft - I made a final commit that slightly adjusted what you had (see inline comment). With that, I think this is ready to go in!!

File: astropy/time/core.py [hunk 2]
```diff
@@ -1622,16 +1712,12 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=False, *, where=True):
         result.format = self.format
         return result
 
-    @property
+    @lazyproperty
     def cache(self):
         """
         Return the cache associated with this instance.
         """
-        return self._time.cache
```
Comment (taldcroft) [inline astropy/time/core.py:1630]: Why this change? It seems a bit out of scope for this PR.

**Review by taldcroft — COMMENTED** [review]
Just working through reading the changes...

File: docs/changes/time/15231.api.rst [hunk 3]
```diff
@@ -0,0 +1,10 @@
+``Time`` has switched to use ``Masked`` arrays internally, instead of
```
[the remark this answers is not in the document]
Comment (mhvk) [reply docs/changes/time/15231.api.rst:1]: I think in the changelog entries, we're not meant to give any links, but we should indeed in whatsnew.

File: docs/whatsnew/6.0.rst [hunk 4]
```diff
@@ -174,6 +175,37 @@ IERS data can now install the latest version of the
 download the package manually and transfer it to a computer that has no
 public internet connection.
 
+.. _whatsnew-6.0-masked-time:
+
+Masked Time values now use Masked arrays internally
+===================================================
+
+|Time| can now be initialized with both |Masked| and `numpy.ma.MaskedArray`
```
[the remark this answers is not in the document]
Comment (mhvk) [reply docs/whatsnew/6.0.rst:247]: Hmm, I didn't think of that. I'll make a change here, but the general issue is worth separating out - see #15516

File: astropy/time/core.py [hunk 5]
```diff
@@ -965,7 +997,8 @@ def to_value(self, format, subfmt="*"):
         cache = self.cache["format"]
         # Try to keep cache behaviour like it was in astropy < 4.0.
         key = format if subfmt is None else (format, subfmt)
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/time/core.py:999]: Yes, makes sense! I think this was me in the first instance wanting to make as little change as possible, and then not thinking enough when adding the configuration option.

File: astropy/time/core.py (hunk 2 above)
Comment (mhvk) [reply astropy/time/core.py:1630]: Just that there is no reason for the cache to stay on the format class - indeed, you originally put it on `Time` itself in #4422; it was moved to the format class to support masking in #6028.

File: astropy/time/core.py [hunk 6]
```diff
@@ -1002,20 +1035,67 @@ def to_value(self, format, subfmt="*"):
...
+        """Get an instance without the mask.
+
+        Note that while one gets a new instance, the underlying data will be shared.
+
+        See Also
+        --------
+        astropy.time.Time.filled
+        """
+        return (
+            self._apply(operator.attrgetter("unmasked"))
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/time/core.py:1078]: Basically, `_apply` applies a function to all array-like attributes and then creates a new instance from the result. Here, the function is to get the `unmasked` attribute.

But looking at this again, this would fail for a location array that is not masked (a possibility I now realize I've happily ignored here - it would work, but the mask would not be reflected in `.mask`... for later??).

To fix this, I've changed this to the following (and added a test):
```
        # Get a new Time instance that has the unmasked versions of all attributes.
        return self._apply(lambda x: getattr(x, "unmasked", x))
```

File: astropy/time/core.py [hunk 7]
```diff
@@ -3237,30 +3313,43 @@ def get_as_filled_ndarray(mask, val):
 
         Note that nothing happens if there are no masked elements.
         """
-        fill_value = None
+        if isinstance(val, np.ma.MaskedArray):
```
[the remark this answers is not in the document]
Comment (mhvk) [reply astropy/time/core.py]: > I think there is a docstring mistake above. It says that the unmasked case returns `mask=None` but it should be `False`. I noticed this in the code that uses this function since it checks for `mask is not False`.

I think you fixed this already yourself, so resolving...

**Comment by mhvk** [comment]
OK, I made the requested changes. The format cache is now much more logical, a bug was found, and an issue raised about the more general problem with the documentation - nice!

File: docs/time/index.rst [hunk 8]
```diff
@@ -854,71 +854,125 @@ Example
...
+Second with the `astropy.utils.masked.Masked` class::
+
+  >>> from astropy.utils.masked import Masked
+  >>> dates = Masked(['2001:020', '', '2001:060'], mask=[False, True, False])
+  >>> t = Time(dates, out_subfmt="date")
+  >>> print(t)
+  ['2001:020'        ——— '2001:060']
+
+An important point is that any value which is marked as missing will be replaced
+internally with a time equivalent to ``2000-01-01 12:00:00`` UTC.
```
Comment (taldcroft) [inline docs/time/index.rst:873]: The time equivalent is not in UTC, it is whatever the scale when creating the object.
```
In [2]: tm = Time(np.ma.array(["..."], mask=[True]), format="isot", scale="tai")

In [3]: tm.unmasked
Out[3]: <Time object: scale='tai' format='isot' value=['2000-01-01T12:00:00.000']>

In [4]: tm.utc.unmasked
Out[4]: <Time object: scale='utc' format='isot' value=['2000-01-01T11:59:28.000']>
```

**Review by taldcroft — COMMENTED** [review]
OK, final review and there is just a tiny documentation issue. Otherwise fantastic @mhvk!

File: docs/time/index.rst (hunk 8 above)
Comment (mhvk) [reply docs/time/index.rst]: Oops, yes, of course, that makes sense since in the end only the value is transferred, not the scale. I adjusted it -- though as I looked at the text, I felt the whole part made more sense as part of the `.. note`. Hope you agree...

**Comment by mhvk** [comment]
p.s. Probably good to check that the `..note` typesets correctly in the docs -- I'm never quite sure what sphinx does with indentation, etc. (and `tox -e build_docs` seems broken locally).

**Comment by mhvk** [comment]
Trying again after removing the empty line between the note and doctest.

**Comment by mhvk** [comment]
OK, the final push worked locally at least...

**Review by taldcroft — APPROVED** [review]
Looks good, approved!!
